### PR TITLE
Update plain to latest version.

### PR DESCRIPTION
**Pull request by le-jzr**
I'm not happy that some unsafe blocks returned, but the API wasn't entirely safe to begin with, so I figure this is better than just sitting on my frustration with the memory model.

Though it would be even better to implement _proper_ reading from files, endianness and all.

### CODE REVIEW

File: src/elf/reloc.rs [hunk 1]
```diff
@@ -191,7 +189,9 @@ macro_rules! elf_rela_std_impl { ($size:ident, $isize:ty) => {
                 let count = size / SIZEOF_RELA;
                 let mut relocs = vec![Rela::default(); count];
                 fd.seek(Start(offset as u64))?;
-                fd.read(relocs.as_mut_bytes())?;
+                unsafe {
+                    fd.read(plain::as_mut_bytes(&mut *relocs))?;
```
Comment (m4b) [inline src/elf/reloc.rs]: The unsafety is technically in the as_mut_bytes so this is slightly over estimated, but I don't care. 

But why is as_mut_bytes unsafe now ?

**Comment by m4b** [comment]
Yea I mean this part of the API is really old, before scroll and etc. I'd be ok with reading the stuff out using scroll but I'm also not too concerned about switching to it. Also there would probably be a perf hit using scroll instead of byte casts

File: src/elf/reloc.rs (hunk 1 above)
Comment (le-jzr) [inline src/elf/reloc.rs]: A bit overestimated, yeah. That's pure aesthetic choice on my part.

The deal with `as_mut_bytes()` is that reading structure padding bytes may be undefined behavior in some cases. It's safe if the structure has no implicit padding, or if you only write without reading, but trying to encode that into the type system is too much trouble.

**Comment by le-jzr** [comment]
I figure I could do a library that uses casting in the case data endianness = host endianness, and `byteorder` crate otherwise. That might be an interesting project. :3

**Comment by m4b** [comment]
Yea that would be ideal, but it's not immediately clear to me how to do runtime switching. Maybe you could hack it by check if a global const is in big endianness or little endianness and if different use scroll (not byteorder :laughing:), and cast otherwise ?

**Comment by le-jzr** [comment]
I figure it would be something like
```
if file.little_endian() {
    file.parse::<Endian::Little>()
} else {
    file.parse::<Endian::Big>()
}
```
but under the hood, the library would instead be specialized for `Endian::Native` vs. `Endian::NonNative`, with `Little` and `Big` being conditional aliases. Or something.

**Comment by m4b** [comment]
I'm not sure how that would work. I think checking a constant's repr similar to macho's FEEDFACE is the easiest way to do it (at least on stable). 

Also the more I think about this the more I want it, but I don't know how it would work. Either you copy (in which case endianness conversions aren't that big a deal (ok for most use cases)) or you get a reference. But a reference to "to be endian converted" struct doesn't make sense. 

Specifically I think it would be really awesome to pread out a reference to a struct. In particular, when I mmap something and I need maximum speed and memory constraints I want to only have references to the backing same endianness struct, and not to copy. 

But I don't see how you could combine the two. 

I would however be interested in putting pread_ref or something into scroll which does do an explicit (but safe) bit cast. 

Maybe call it reinterpret a la c++ but safe?

**Pull request merged**